### PR TITLE
Add filter function for NeuralQueryBuilder and HybridQueryBuilder and…

### PR DESCRIPTION
**Pull request by chloewqg**
… modify fromXContent function in HybridQueryBuilder to support filter field.

### Description
Add filter function to NeuralQueryBuilder and HybridQueryBuilder, which allows to push down non null filter. One exception is that when HybridQueryBuilder has a nested HybridQueryBuilder, then calling the filter function will cause UnsupportedOperationException

### Related Issues
Resolves #1206 
Related: #282, #1135

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java [hunk 1]
```diff
@@ -309,6 +309,23 @@ protected void doWriteTo(StreamOutput out) throws IOException {
         RescoreParser.streamOutput(out, rescoreContext);
     }
 
+    /**
+     * Add a filter to Neural Query Builder
+     * @param filterToBeAdded filter to be added
+     * @return return itself with underlying filter combined with passed in filter
+     */
+    public QueryBuilder filter(QueryBuilder filterToBeAdded) {
+        if (validateFilterParams(filterToBeAdded) == false) {
```
Comment (junqiu-lei) [inline src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java:318]: the method `validateFilterParams` is missed i think.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java [hunk 2]
```diff
@@ -94,6 +95,31 @@ public HybridQueryBuilder add(QueryBuilder queryBuilder) {
...
+     * and throw exception telling this is an unsupported operation. If the filter is null, then we do nothing and
+     * return.
+     * @param filter the filter parameter
+     * @return HybridQueryBuilder itself
+     */
+    public QueryBuilder filter(QueryBuilder filter) {
+        if (!validateFilterParams(filter)) {
+            return this;
+        }
+        HybridQueryBuilder compoundQueryBuilder = new HybridQueryBuilder();
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: Why do we have to create a new query builder instead of returning `this`?

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: I cannot find validateFilterParams method. Besides that the naming is confusing, if you conclude validity of the object based on return value that needs to be in the name, something like `isValidFilterParams`. Also change negation to an explicit == false

**Review by martin-gaievski — COMMENTED** [review]
there is a separate issue and the whole design/RFC for hybrid query. Is this a building block for that design or this is a conflicting issue? 
Ref: https://github.com/opensearch-project/neural-search/issues/282, https://github.com/opensearch-project/neural-search/issues/1135

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 2 above)
Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: Hi the validateFilterParams is in the core package, inside the AbstractQueryBuilder class. https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/AbstractQueryBuilder.java#L94

Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: > Why do we have to create a new query builder instead of returning `this`?

Because I saw the the queries of HybridQueryBuilder is marked as final. https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java#L56

So I think as we need to reassign a list of queries to the queries variable inside the HybridQueryBuilder. So we can only create a new HybridQueryBuilder to solve this?

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java [hunk 3]
```diff
@@ -94,6 +95,31 @@ public HybridQueryBuilder add(QueryBuilder queryBuilder) {
         return this;
     }
 
+    /**
+     * Function to support filter function on HybridQueryBuilder filter. Currently pushing down a filter
```
[the remark this answers is not in the document]
Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:99]: Updated.

**Comment by chloewqg** [comment]
> there is a separate issue and the whole design/RFC for hybrid query. Is this a building block for that design or this is a conflicting issue? Ref: #282, #1135

Hi Martin, this is a building block for that design. cc @bzhangam

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 2 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: > Thanks Martin for the comment! Actually the function name `validateFilterParams` was suggested by Daniel in the pull request of core package. [opensearch-project/OpenSearch#17409](https://github.com/opensearch-project/OpenSearch/pull/17409) , is it okay to leave as what it is for now?

if method is already there let's utilize it. Please use == false syntax instead of negation, I see you've used it in other places in your PR

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: > > Why do we have to create a new query builder instead of returning `this`?
> 
> Because I saw the the queries of HybridQueryBuilder is marked as final. https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java#L56
> 
> So I think as we need to reassign a list of queries to the queries variable inside the HybridQueryBuilder. So we can only create a new HybridQueryBuilder to solve this?

Can't we do this.add(query.filter(filter))? Final doesn't mean we cannot mutate the object, just cannot point ref to another object

**Comment by martin-gaievski** [comment]
> > there is a separate issue and the whole design/RFC for hybrid query. Is this a building block for that design or this is a conflicting issue? Ref: #282, #1135
> 
> Hi Martin, this is a building block for that design. cc @bzhangam

Got it, makes sense then. Can you please mention same in the PR description and also update the link to the feature request issue under "Related issues"? If the filtering feature requires multiple PRs/building blocks I suggest you create a high-level meta issue and host smaller PRs under that meta issue.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 2 above)
Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: Wouldn't this double the filters sizes? Like if there are two filters, when we add(query.filter(filter)), wouldn't this create duplicate filters?

Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: > > Thanks Martin for the comment! Actually the function name `validateFilterParams` was suggested by Daniel in the pull request of core package. [opensearch-project/OpenSearch#17409](https://github.com/opensearch-project/OpenSearch/pull/17409) , is it okay to leave as what it is for now?
> 
> if method is already there let's utilize it. Please use == false syntax instead of negation, I see you've used it in other places in your PR

Yep. Updated in the newer commit

**Comment by chloewqg** [comment]
> > > there is a separate issue and the whole design/RFC for hybrid query. Is this a building block for that design or this is a conflicting issue? Ref: #282, #1135
> > 
> > 
> > Hi Martin, this is a building block for that design. cc @bzhangam
> 
> Got it, makes sense then. Can you please mention same in the PR description and also update the link to the feature request issue under "Related issues"? If the filtering feature requires multiple PRs/building blocks I suggest you create a high-level meta issue and host smaller PRs under that meta issue.

Gotcha. Thanks!

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 2 above)
Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: > Quote reply

I think the filter input check is common across all the query builders. Although current check is simple of whether it equals to null, adding a function here can allow complicated logic changes in the future

Comment (bzhangam) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: The reason we need to create a new HybridQueryBuilder here is because this function is a static function and we cannot set the value for the parameters in a static function unless we also make them static. I don't think we want to make all the parameters static so we have to create an instance in this static function.

File: src/test/java/org/opensearch/neuralsearch/query/NeuralQueryIT.java [hunk 4]
```diff
@@ -176,6 +177,101 @@ public void testQueryWithBoostAndImageQueryAndRadialQuery() {
...
+     *                          "match_none" : {}
+     *                     ]
+     *                 }
+     *             }
+     *         }
+     *     }
+     * }
+     */
+    @SneakyThrows
+    public void testQueryWithBoostAndFilterApplied() {
```
Comment (bzhangam) [inline src/test/java/org/opensearch/neuralsearch/query/NeuralQueryIT.java:244]: Why we add this IT? I think the filter function is an existing feature for neural query?

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java [hunk 5]
```diff
@@ -167,6 +174,93 @@ private void testPostFilterRangeQuery(String indexName) {
...
+                                }
+                            }
+                        ]
+                    }
+                ]
+            }
+        }
+    }*/
+    @SneakyThrows
+    private void testRangeQueryAsFilter(String indexName) {
```
Comment (bzhangam) [inline src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java:235]: Why we add this IT in PostFilterIT? Filter is different from post filter. I think we should either add it to the HybridQueryIT or create a new HybridQueryFilterIT.

Comment (bzhangam) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java:235]: Do you plan to raise another PR to add tests for BWC? BWC tests are under the qa folder.

Comment (martin-gaievski) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java:235]: > Why we add this IT in PostFilterIT? Filter is different from post filter. I think we should either add it to the HybridQueryIT or create a new HybridQueryFilterIT.

+1, better add new class HybridQueryFilterIT

Comment (chloewqg) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java:235]: > Do you plan to raise another PR to add tests for BWC? BWC tests are under the qa folder.

I think the current bwc tests is blocking the merge. So I will do the bwc tests here in the same PR

Comment (chloewqg) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java:235]: > Why we add this IT in PostFilterIT? Filter is different from post filter. I think we should either add it to the HybridQueryIT or create a new HybridQueryFilterIT.

Sure. Will create a new HybridQueryFilterIT

File: src/test/java/org/opensearch/neuralsearch/query/NeuralQueryIT.java (hunk 4 above)
Comment (chloewqg) [reply src/test/java/org/opensearch/neuralsearch/query/NeuralQueryIT.java:244]: Here we are testing whether adding another filter on NeuralQuery would be function as expected.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java [hunk 6]
```diff
@@ -94,6 +95,28 @@ public HybridQueryBuilder add(QueryBuilder queryBuilder) {
...
+     * and throw exception telling this is an unsupported operation. If the filter is null, then we do nothing and
+     * return.
+     * @param filter the filter parameter
+     * @return HybridQueryBuilder itself
+     */
+    public QueryBuilder filter(QueryBuilder filter) {
+        if (validateFilterParams(filter) == false) {
+            return this;
+        }
+        for (int i = 0; i < queries.size(); i++) {
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: This is fine because the queries are array list. However, we should avoid the assumption that it will be ArrayList especially because we defined as `List`. 
```
private final List<QueryBuilder> queries = new ArrayList<>();
```

Therefore, following is better so that its performance does not degrade even if we switch to other list type.
```
ListIterator<QueryBuilder> iterator = queries.listIterator();
while(iterator.hasNext()) {
  QueryBuilder query = iterator.next();
  if (query instanceof HybridQueryBuilder) {
      throw new UnsupportedOperationException("Cannot push filter to nested hybridQueryBuilder");
  }
  iterator.set(query.filter(filter));
}
```

However, on a second thought, why do we need to set the query again? Also, we don't need to check if query is type of HybridQuery or not. It should be checked somewhere else. So we can simple do like
```
queries.stream().forEach(query -> query.filter(filter));
```

Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:110]: Gotcha! Thanks Heemin! Btw, I am little confused in the current bwc tests failure. Seems like the error message is 

```
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: 127.0.0.1/127.0.0.1:42235
»  Caused by: java.net.ConnectException: Connection refused
```

I am bit confused on what is wrong here with my filter function changes.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java [hunk 7]
```diff
@@ -94,6 +96,30 @@ public HybridQueryBuilder add(QueryBuilder queryBuilder) {
...
+     * @return HybridQueryBuilder itself
+     */
+    public QueryBuilder filter(QueryBuilder filter) {
+        if (validateFilterParams(filter) == false) {
+            return this;
+        }
+        ListIterator<QueryBuilder> iterator = queries.listIterator();
+        while(iterator.hasNext()) {
+            QueryBuilder query = iterator.next();
+            if (query instanceof HybridQueryBuilder) {
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:114]: I don't think we need this check here?

Comment (heemin32) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:114]: That will be handled by https://github.com/opensearch-project/neural-search/pull/498/files#diff-64ab60781cf1e9f7963a415406f49283b26945beaa46b7042e7cc5044d8d95b1R74

I don't think this method should worry about it.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java [hunk 8]
```diff
@@ -94,6 +96,27 @@ public HybridQueryBuilder add(QueryBuilder queryBuilder) {
         return this;
     }
 
+    /**
+     * Function to support filter on HybridQueryBuilder filter. Currently pushing down a filter
+     * to HybridQueryBuilder is not supported by design. We would simply check if the filter is valid
+     * and throw exception telling this is an unsupported operation. If the filter is null, then we do nothing and
```
Comment (bzhangam) [inline src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:102]: The description doesn't match the implementation. Have we changed the mind that we want to support the filter push down for a hybrid query? Basically it means we want to support a hybrid query nested in another hybrid query. I don't think we plan to support it in near future.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java [hunk 9]
```diff
@@ -51,6 +52,7 @@ public final class HybridQueryBuilder extends AbstractQueryBuilder<HybridQueryBu
     public static final String NAME = "hybrid";
 
     private static final ParseField QUERIES_FIELD = new ParseField("queries");
+    private static final ParseField FILTER_FIELD = new ParseField("filter");
```
Comment (bzhangam) [inline src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:55]: The new field should also be added to **HybridQueryBuilder(StreamInput in)** and **doXContent(XContentBuilder builder, Params params)** so that we will lose the filter info when we pass it across nodes. Even though the filter info is already included in the sub queries I think we still should follow the best practice to include it in those two functions.

Besides we also want to add it to the doEquals and doHashCode function in case they are used to compare two hybrid queries.

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java (hunk 5 above)
Comment (bzhangam) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java:235]: I saw you created HybridQueryFilterIT. Do you plan to remove this test?

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 8 above)
Comment (heemin32) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:102]: I asked to remove the check of nested hybrid query here. I believe the blocking of the nested hybrid query should be handled elsewhere, but it doesn’t necessarily have to be in this specific location.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 9 above)
Comment (heemin32) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:55]: It seems like we are missing some test case here unless this filter setting is happening in coordination node.

File: src/test/java/org/opensearch/neuralsearch/query/NeuralQueryIT.java (hunk 4 above)
Comment (bzhangam) [reply src/test/java/org/opensearch/neuralsearch/query/NeuralQueryIT.java:244]: I think IT is mock a real use case. If we want to test that we should test a hybrid query with a neural sub query. And both the hybrid query and the neural query have a filter.

In this test we simply test the filter function of the neural query builder which has already be done as a unit test.

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+                                }
+                            }
+                        ]
+                    }
+                ]
+            }
+        }
+    }*/
+    @SneakyThrows
+    private void testRangeQueryAsFilter(String indexName) {
```
Comment (bzhangam) [inline src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java:130]: I think most common use case for hybrid query is using a neural query and a normal term query as sub queries. Can we build the IT for that use case?

One important thing we want to validate is that for neural query we should push the filter into it. It can give us different result compared to the case we use a bool query to combine a filter with a neural query. This happens because the filter in neural query will be applied before we return the top k result while the filter combined through a bool query will happen after returning the top k result. Would recommend to build a test case for that.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 8 above)
Comment (heemin32) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:102]: Still I think we should update the comment which does not match with the actual implementation.

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java (hunk 10 above)
Comment (chloewqg) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java:130]: Okay, sounds good

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 9 above)
Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:55]: Sounds good

**Review by vibrantvarun — COMMENTED** [review]
1. Please add bwc tests as this field is added new the query. We need to evaluate the impact in case of bwc environment.
2. Why we have not added this field in HybridQuery.java?

cc: @martin-gaievski @heemin32

**Comment by heemin32** [comment]
> 2. Why we have not added this field in HybridQuery.java?
We just pass the filter down to sub queries. Therefore, there is no need for hybrid query to hold it as we are not going to filter after the sub queries return the result.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 9 above)
Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:55]: Hi Bo, quick question while I am trying to follow your comments. 
```
The new field should also be added to HybridQueryBuilder(StreamInput in) and doXContent(XContentBuilder builder, Params params) so that we will lose the filter info.
```
What do you mean by we will lose the filter info? Could you please give an example?

```
add it to the doEquals and doHashCode function in case they are used to compare two hybrid queries.
```

The field I added here is not the filter itself but the filter parser field name. I believe the filter itself is pushed down to queries. So we are automatically comparing the queries and its filter in doEquals and doHashCode. What do you mean by adding it to the doEquals and doHashCode function?

Comment (bzhangam) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:55]: For your first question when we pass the query across nodes we will convert it to a stream and later in another node build it based on the stream. If we don't clearly write the filter to the stream we will lose it.

For the second question we override the doEquals and doHashCode function so if we don't clearly check it those two functions will ignore it.

Comment (bzhangam) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:55]: Synced offline. Since currently we always want to push down the filter to sub-queries and we always do this work in the fromXContent function there is no need to persist the filter in the HybridQuery. In this case there is not need to write it to the stream since it's already included in the sub-queries. Same for the doEqual and doHash functions.

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:55]: I think this should be fine, one concern I do have is some edge case in context of cluster upgrade. Can you add a bwc test, it may be skipped for now for 2.x -> 3.x migration, but it will run in future for next 3.x, e.g. 3.0 -> 3.1. You can check how it's done in of the recent PR for stats API

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 8 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:102]: Please state in the comment that filter is not stored in the top hybrid query but only pushed down to individual sub-queries

File: src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java [hunk 11]
```diff
@@ -309,6 +309,23 @@ protected void doWriteTo(StreamOutput out) throws IOException {
...
+            return this;
+        }
+        if (filter == null) {
+            filter = filterToBeAdded;
+            return this;
+        }
+        filter = filter.filter(filterToBeAdded);
+        return this;
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java:326]:
```suggestion
        if (filter == null) {
            filter = filterToBeAdded;
        } else {
             filter = filter.filter(filterToBeAdded);
        }
        return this;
```

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 8 above)
Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:102]: Yes. Will publish new revision for that.

File: src/test/java/org/opensearch/neuralsearch/query/NeuralQueryIT.java [hunk 12]
```diff
@@ -176,6 +177,101 @@ public void testQueryWithBoostAndImageQueryAndRadialQuery() {
         );
     }
 
+    /**
+     * Tests basic query without filter:
```
[the remark this answers is not in the document]
Comment (chloewqg) [reply src/test/java/org/opensearch/neuralsearch/query/NeuralQueryIT.java:181]: Okay, I will remove the test here.

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java [hunk 13]
```diff
@@ -74,6 +74,13 @@ public void testPostFilterOnIndexWithSingleShard_whenConcurrentSearchEnabled_the
         testPostFilterMatchAllAndMatchNoneQueries(TEST_MULTI_DOC_INDEX_WITH_TEXT_AND_INT_SINGLE_SHARD);
     }
 
+    @SneakyThrows
+    public void testFilterOnIndexWithSingleShard_whenConcurrentSearchEnabled_thenSuccessful() {
```
[the remark this answers is not in the document]
Comment (chloewqg) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryPostFilterIT.java:78]: Yes will remove it.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java [hunk 14]
```diff
@@ -94,6 +96,27 @@ public HybridQueryBuilder add(QueryBuilder queryBuilder) {
...
+     */
+    public QueryBuilder filter(QueryBuilder filter) {
+        if (validateFilterParams(filter) == false) {
+            return this;
+        }
+        ListIterator<QueryBuilder> iterator = queries.listIterator();
+        while (iterator.hasNext()) {
+            QueryBuilder query = iterator.next();
+            // set the query again because query.filter(filter) can return new query.
+            iterator.set(query.filter(filter));
```
[the remark this answers is not in the document]
Comment (chloewqg) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:115]: Sounds good

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java [hunk 15]
```diff
@@ -94,6 +96,29 @@ public HybridQueryBuilder add(QueryBuilder queryBuilder) {
...
+     */
+    public QueryBuilder filter(QueryBuilder filter) {
+        if (validateFilterParams(filter) == false) {
+            return this;
+        }
+        ListIterator<QueryBuilder> iterator = queries.listIterator();
+        while (iterator.hasNext()) {
+            final QueryBuilder query = iterator.next();
+            final QueryBuilder queryWithFilter = query.filter(filter);
+            if (!queryWithFilter.equals(query)) {
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:114]: don't use `equals` method, you need to compare exact reference, like `queryWithFilter == query`. Also use == false instead of negation in if condition

File: qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/restart/HybridSearchWithRescoreIT.java [hunk 16]
```diff
@@ -64,10 +64,15 @@ public void testHybridQueryWithRescore_whenIndexWithMultipleShards_E2EFlow() thr
...
-                HybridQueryBuilder hybridQueryBuilder = getQueryBuilder(modelId, null, null);
+                HybridQueryBuilder hybridQueryBuilder = getQueryBuilder(modelId, null, null, null);
                 QueryBuilder rescorer = QueryBuilders.matchQuery(TEST_FIELD, RESCORE_QUERY).boost(0.3f);
                 validateTestIndex(getIndexNameForTest(), hybridQueryBuilder, rescorer);
-                hybridQueryBuilder = getQueryBuilder(modelId, Map.of("ef_search", 100), RescoreContext.getDefault());
+                hybridQueryBuilder = getQueryBuilder(
+                    modelId,
+                    Map.of("ef_search", 100),
+                    RescoreContext.getDefault(),
+                    new MatchQueryBuilder("_id", "1")
```
Comment (martin-gaievski) [inline qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/restart/HybridSearchWithRescoreIT.java:74]: why this change is needed, it doesn't look like a simple adoption of the new argument?

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+
+        Map<String, Object> searchResponseAsMapTextQuery = search(indexName, hybridQueryBuilder, 1);
+        assertEquals(1, getHitCount(searchResponseAsMapTextQuery));
+
+        Map<String, Object> firstInnerHitTextQuery = getFirstInnerHit(searchResponseAsMapTextQuery);
+        assertEquals("1", firstInnerHitTextQuery.get("_id"));
+    }
+
+    @SneakyThrows
+    public void testFilterOnIndexWithSingleShard_whenConcurrentSearchEnabled_thenSuccessful() {
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java:106]: I don't think we need this match of tests for CSS, one should be enough, this one you can delete

File: qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/restart/HybridSearchWithRescoreIT.java (hunk 16 above)
Comment (chloewqg) [reply qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/restart/HybridSearchWithRescoreIT.java:74]: Oh I checked previous PR and found they modify this for bwc tests. Shall I discard this change?

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+
+    @SneakyThrows
+    public void testFilterOnNeuralQueryFilterAndTermQueryFilter_thenSuccessful() {
+        updateClusterSettings(CONCURRENT_SEGMENT_SEARCH_ENABLED, true);
+        prepareResourcesBeforeTestExecution(SHARDS_COUNT_IN_SINGLE_NODE_CLUSTER);
+        testNeuralQueryBuilder(TEST_MULTI_DOC_INDEX_WITH_TEXT_AND_INT_SINGLE_SHARD);
+    }
+
+    @SneakyThrows
+    private void testNeuralQueryBuilder(String indexName) {
```
[the remark this answers is not in the document]
Comment (chloewqg) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java:80]: What do you mean by set the CSS setting explicitly?

[the remark this answers is not in the document]
Comment (chloewqg) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java:80]: Another question for [Build and Test Neural Search / Pre-commit Linux (23, ubuntu-latest) ] Failure. It is caused by `java.lang.NoClassDefFoundError: Could not initialize class org.opensearch.knn.index.engine.KNNEngine` Yesterday it works fine offline. What might be causing this issue?

[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java:80]: > What do you mean by set the CSS setting explicitly?

if you want CSS to be disabled for the test, make sure to explicitly set it to `false`. It looks like you're enabling it in tests where you need it, and assuming that it will be reset to `false` before each next test. It's not the case if tests are sharing same index. In other words your test may be running against index where CSS is enabled when it's not required

File: qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/restart/HybridSearchWithRescoreIT.java (hunk 16 above)
Comment (martin-gaievski) [reply qa/restart-upgrade/src/test/java/org/opensearch/neuralsearch/bwc/restart/HybridSearchWithRescoreIT.java:74]: if by default filter can be `null` then pass it as `null`. With current logic you're testing multiple things, but the test it built just for rescore functionality

File: src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java:80]: > Another question for [Build and Test Neural Search / Pre-commit Linux (23, ubuntu-latest) ] Failure. It is caused by `java.lang.NoClassDefFoundError: Could not initialize class org.opensearch.knn.index.engine.KNNEngine` Yesterday it works fine offline. What might be causing this issue?

which run do you mean? I checked few failed ones at random, they are all flaky, model haven't loaded in time:

```
java.lang.AssertionError: failed to load the model, last task finished with status RUNNING
```

[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/test/java/org/opensearch/neuralsearch/query/HybridQueryFilterIT.java:80]: Another failure related to some recent changes in core, I think it's covered by this issue: https://github.com/opensearch-project/neural-search/issues/1232

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 15 above)
Comment (heemin32) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:114]: @martin-gaievski I don’t see any benefit from this change. The previous code was much better—using a `set` isn’t costly, even compared to an `equals` check. The earlier implementation was also simpler and more readable. Would it be possible to revert to the previous version?

By the way, in this code, we are calling `query.filter(filter);` two times.

File: src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (heemin32) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:114]: @martin-gaievski, If you're okay with it, let's keep this as is. The difference between `equals` and `set` is negligible, so a simpler approach would be preferable.

[the remark this answers is not in the document]
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java:114]: ok, let's keep the `set()` method, I agree that original version of the code is simpler

**Review by martin-gaievski — APPROVED** [review]
